cmd/kubeadm/app/cmd/defaults.go

```diff
@@ -35,14 +35,14 @@ var (
 )
 
 func setInitDynamicDefaults(cfg *kubeadmapi.MasterConfiguration) error {
-	// Auto-detect the IP
-	if len(cfg.API.AdvertiseAddresses) == 0 {
-		ip, err := netutil.ChooseHostInterface()
-		if err != nil {
-			return err
-		}
-		cfg.API.AdvertiseAddresses = []string{ip.String()}
+
+	// Choose the right bind address. If the bind address is localhost or 0.0.0.0, the default interface's IP address is used
+	// This is the same logic as the API Server uses
+	ip, err := netutil.ChooseBindAddress(net.ParseIP(cfg.API.BindAddress))
+	if err != nil {
+		return err
 	}
+	cfg.API.BindAddress = ip.String()
```

Comment:
This is wrong.  If you start k8s with a bind address of 0.0.0.0 it will *listen* on all interfaces but will only advertise (as in set in certs and such) the external interface.

The right thing to do here is to have bind address and PublicAddress be 2 different options and have `netutil.ChooseBindAddress` be used to set PublicAddress.  (`netutil.ChooseBindAddress` is a misnomer as isn't used that way).

While 1.5 worked this way, this may break 1.6 as we are no longer opening up the insecure port on localhost.